File: backend/src/routes/cities.js
import express from 'express';
import { prisma } from '../db.js';
import { CG_CITIES } from '../config.js';
import { get as cacheGet, set as cacheSet } from '../cache.js';

const router = express.Router();

/** GET /api/cities — 33 CG districts + open-tender counts, plus "Unspecified" */
router.get('/', async (_req, res) => {
  try {
    const cached = cacheGet('cities');
    if (cached) {
      return res.json(cached);
    }

    const grouped = await prisma.tender.groupBy({
      by: ['locationCity'],
      where: { status: 'open' },
      _count: { _all: true },
    });

    const countMap = Object.fromEntries(grouped.map((g) => [g.locationCity, g._count._all]));

    const cities = CG_CITIES.map((name) => ({
      name,
      slug: name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, ''),
      openCount: countMap[name] || 0,
    }));

    cities.push({
      name: 'Unspecified',
      slug: 'unspecified',
      openCount: countMap['Unspecified'] || 0,
    });

    const result = { cities };
    cacheSet('cities', result, 43200000); // 12 hours

    res.json(result);
  } catch (e) {
    console.error('[api] GET /cities error:', e);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend/app/routes/cities.py
"""
app/routes/cities.py

Direct port of src/routes/cities.js.
"""
import re
from datetime import datetime, timezone

from fastapi import APIRouter
from fastapi.responses import JSONResponse

from .. import db
from ..config import CG_CITIES
from ..cache import get as cache_get, set as cache_set

router = APIRouter(prefix="/api/cities", tags=["cities"])


@router.get("")
@router.get("/")
async def list_cities():
    """GET /api/cities — 33 CG districts + open-tender counts, plus "Unspecified" """
    try:
        cached = cache_get('cities')
        if cached:
            return JSONResponse(cached)

        now = datetime.now(timezone.utc)
        pipeline = [
            {'$match': {
                'status': 'open',
                '$or': [{'endDate': {'$gte': now}}, {'endDate': None}],
            }},
            {'$group': {'_id': '$locationCity', 'count': {'$sum': 1}}},
        ]
        grouped = await db.tenders.aggregate(pipeline).to_list(length=None)
        count_map = {g['_id']: g['count'] for g in grouped}

        def slugify(name: str) -> str:
            s = re.sub(r'[^a-z0-9]+', '-', name.lower())
            return s.strip('-')

        cities = [
            {'name': name, 'slug': slugify(name), 'openCount': count_map.get(name, 0)}
            for name in CG_CITIES
        ]
        cities.append({'name': 'Unspecified', 'slug': 'unspecified', 'openCount': count_map.get('Unspecified', 0)})

        result = {'cities': cities}
        cache_set('cities', result, 43200000)  # 12 hours

        return JSONResponse(result)
    except Exception as e:
        print(f"[api] GET /cities error: {e}")
        return JSONResponse({'error': 'Internal server error'}, status_code=500)
